### DISCUSSION TITLE
Play by Email (PBEM)

### DISCUSSION THREAD

**Discussion by StefanDK** [bafolts/terraforming-mars#4748] [Ideas]
It could be awesome to play this game by email. 

I have looked into building the source code and hosting the server and that is easy enough.

I am not too familiar with Node.js, but it seems quite easy to implement sending emails when a player's turn is up - like a notification. There could be a few settings for the server that sets the correct SMTP information - and then it would be up to the server host to provide the email info.

When logging in to the game for the first time or maybe even when creating the game (number of players, settings etc.) the player could be asked to give an email to be contacted by.

Then the game could send emails when something important happens (change of turns/Mars terraformed etc.) and the game could be played over a few days.

I will try to look into this myself at some point.

This could also be heavily utilised in tournament setups where it can be difficult to coordinate many players.

**Comment by bafolts** [comment]
If we made some place to hook these events then adding the email notification should be straight forward. I would be hesitant to store people's emails from privacy perspective by default. If there were a hook of _when_ the notification had to be sent then someone would have to listen for that event and send the emails. Could be other use cases too besides email.

**Reply by StefanDK** [reply]
Yes, GDPR and stuff like that would mean that there should be no persisting the email address at all. But holding the data in memory or by cookie or something like that should be possible.

In the code I searched the " ... it is your turn ..." (or similar) and found code that managed the notification. If the email and smtp info is available then it should be pretty straightforward. 

But there would of course be a measurable amount of work as regards the UI, logic and setting the SMTP info. 

As said, I might look into this myself at some point, but I do not have Node.js experience and have lots of these kinds of projects (primarily outside GitHub).

**Reply by StefanDK** [reply]
I have started looking into this again. I was thinking of trying to implement push notifications using the browser. This would obviously need to be stored in the database or in a file or likewise, but it would not contain anything than a subscription object identifying the user and not an email.

The subscription object would need to contain a list og playerIds so that the player can be notified when it's their turn.

I have forked the repo and made a branch on my fork and will see if I can implement it. If nothing else, just as a prototype.

Feel free to comment ...
